src/cunumeric/stat/histogram.cuh

```diff
@@ -0,0 +1,148 @@
+
+#include <thrust/binary_search.h>
+#include <thrust/device_vector.h>
+#include <thrust/fill.h>
+#include <thrust/iterator/constant_iterator.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/sort.h>
+
+#include <thrust/reduce.h>
+#include <thrust/scatter.h>
+
+#include <thrust/execution_policy.h>
+#include <thrust/system/omp/execution_policy.h>
+#include <cuda_runtime.h>
+#include <thrust/version.h>
+
+#include <cub/cub.cuh>
+
+#ifdef _DEBUG
+#include <cassert>
+#include <iostream>
+#endif
+
+#include "cunumeric/utilities/thrust_util.h"
+
+#include "cunumeric/stat/histogram_gen.h"
+
+namespace cunumeric {
+namespace detail {
+template <typename element_t>
+decltype(auto) get_raw_ptr(Buffer<element_t>& v)
+{
+  return v.ptr(0);
+}
+
+// rudimentary device allocator:
+//
+template <typename elem_t, typename exe_policy_t>
+struct allocator_t<
+  elem_t,
+  exe_policy_t,
+  std::enable_if_t<!std::is_same_v<exe_policy_t, thrust::detail::host_t> &&
+                   !std::is_same_v<exe_policy_t, thrust::system::omp::detail::par_t>>> {
+  allocator_t(void) {}
+
+  allocator_t(elem_t, exe_policy_t) {}  // tag-dispatch for CTAD
+
+  elem_t* operator()(exe_policy_t exe_pol, size_t size, elem_t init = 0)
+  {
+    d_buffer_     = create_buffer<elem_t>(size, Legion::Memory::Kind::GPU_FB_MEM);
+    elem_t* d_ptr = get_raw_ptr(d_buffer_);
+
+    thrust::fill_n(exe_pol, d_ptr, size, init);
+
+    return d_ptr;
+  }
+
+ private:
+  Buffer<elem_t> d_buffer_;
+};
+
+template <typename exe_policy_t>
+void synchronize_exec(exe_policy_t, cudaStream_t stream)
+{
+  CHECK_CUDA(cudaStreamSynchronize(stream));
+}
+
+// device version:
+//
+template <typename exe_policy_t, typename weight_t, typename offset_t, typename allocator_t>
+struct segmented_sum_t<
+  exe_policy_t,
+  weight_t,
+  offset_t,
+  allocator_t,
+  std::enable_if_t<!std::is_same_v<exe_policy_t, thrust::detail::host_t> &&
+                   !std::is_same_v<exe_policy_t, thrust::system::omp::detail::par_t>>> {
+  segmented_sum_t(exe_policy_t exe_pol,
+                  weight_t const* p_weights,
+                  size_t n_samples,
+                  weight_t* p_hist,
+                  size_t n_intervals,
+                  offset_t* p_offsets,
+                  cudaStream_t stream,
+                  allocator_t& allocator)
+    : ptr_weights_(p_weights),
+      n_samples_(n_samples),
+      ptr_hist_(p_hist),
+      n_intervals_(n_intervals),
+      ptr_offsets_(p_offsets),
+      stream_(stream)
+  {
+    cub::DeviceSegmentedReduce::Sum(p_temp_storage_,
+                                    temp_storage_bytes_,
+                                    ptr_weights_,
+                                    ptr_hist_,
+                                    n_intervals_,
+                                    ptr_offsets_,
+                                    ptr_offsets_ + 1,
+                                    stream_);
+
+    p_temp_storage_ = allocator(exe_pol, temp_storage_bytes_);
```

Comment:
No, b/c I'm using `allocator` in more places that just the `segmented_sum_t`. It's also used in `histogram_weights()` (impl).

Reply:
I'm not suggesting changing the definition of the allocator type itself, just making `alloc_scratch_` a member variable of the `segmented_sum_t` class, rather than allocating it externally and passing it in through *shudder* non-const lvalue.

```
template <typename exe_policy_t, typename weight_t, typename offset_t>
struct segmented_sum_t<
  exe_policy_t,
  weight_t,
  offset_t,
  std::enable_if_t<!std::is_same_v<exe_policy_t, thrust::detail::host_t> &&
                   !std::is_same_v<exe_policy_t, thrust::system::omp::detail::par_t>>> {
  segmented_sum_t(exe_policy_t exe_pol,
                  weight_t const* p_weights,
                  size_t n_samples,
                  weight_t* p_hist,
                  size_t n_intervals,
                  offset_t* p_offsets,
                  cudaStream_t stream)
    : ptr_weights_(p_weights),
      n_samples_(n_samples),
      ptr_hist_(p_hist),
      n_intervals_(n_intervals),
      ptr_offsets_(p_offsets),
      stream_(stream)
  {
    cub::DeviceSegmentedReduce::Sum(p_temp_storage_,
                                    temp_storage_bytes_,
                                    ptr_weights_,
                                    ptr_hist_,
                                    n_intervals_,
                                    ptr_offsets_,
                                    ptr_offsets_ + 1,
                                    stream_);

    p_temp_storage_ = alloc_scratch_(exe_pol, temp_storage_bytes_);
  }

  void operator()(void)
  {
    cub::DeviceSegmentedReduce::Sum(p_temp_storage_,
                                    temp_storage_bytes_,
                                    ptr_weights_,
                                    ptr_hist_,
                                    n_intervals_,
                                    ptr_offsets_,
                                    ptr_offsets_ + 1,
                                    stream_);
  }

  weight_t* get_histogram(void) { return ptr_hist_; }

 private:
  weight_t const* ptr_weights_{nullptr};
  size_t n_samples_{0};
  weight_t* ptr_hist_{nullptr};
  size_t n_intervals_{0};
  offset_t* ptr_offsets_{nullptr};
  cudaStream_t stream_{nullptr};
  allocator_t<unsigned char, exe_policy_t> alloc_scratch_{};
  void* p_temp_storage_{nullptr};
  size_t temp_storage_bytes_{0};
};
```

Reply:
Okay, done.

For the record, though: removing the `alloc_t` template parameter from `segmented_sum_t` template is not ideal, because now the allocator is hardwired to `allocator_t` type. What if I want to be able (say, for debugging reasons) to change / experiment w/ a different type of allocator, inside `histogram.cuh: segmented_sum_t`, etc.